testing/upload_validation.ts

```diff
@@ -862,9 +862,28 @@ const syncTableSchema = z.union([genericDynamicSyncTableSchema, genericSyncTable
 // (Zod doesn't let you call .extends() after you've called .refine(), so we're only refining the top-level
 // schema we actually use.)
 const unrefinedPackVersionMetadataSchema = zodCompleteObject<PackVersionMetadata>({
-  version: z
-    .string()
-    .regex(/^\d+(\.\d+){0,2}$/, 'Pack versions must use semantic versioning, e.g. "1", "1.0" or "1.0.0".'),
+  version: z.string().refine(
+    version => {
+      const match = version.match(/^(\d+)(\.\d+)?(\.\d+)?$/);
```

Comment:
Would it work to just change each of these from `\d+` to, say, `\d{1,10}` or whatever a compatible character limit is? (Acknowledging that we'll wind up disallowing some large castable values.)

Alternatively, can you just add a `refine()` after the `regex()` call, and just call `split('.')` on it since we know it's syntactically valid at that point?

`return value.split('.').all(num => Number(num) <= ...);`

And doing the refine separately allows for an error message specific to this case.

Reply:
Great idea to split it up. Done!